### ISSUE TITLE
Casting of the date/timestamps to smaller types should return min/max instead of overflow

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#39249]
```sql
SELECT
    toDate32('2160-01-01') AS a,
    toDate('2160-01-01') AS b,
    toDate(a),
    toDateTime(a),
    toDateTime(b)

Query id: b1043ea4-a5c4-4ab8-98f2-e6d0fbcfbdf5

Row 1:
──────
a:                                  2160-01-01
b:                                  2149-06-06
toDate(toDate32('2160-01-01')):     1980-07-27 -- max date value expected, i.e. 2149-06-06  
toDateTime(toDate32('2160-01-01')): 2023-11-24 17:31:44 -- max datetime value expected, i.e. 2106-02-07 06:28:15
toDateTime(toDate('2160-01-01')):   2013-04-29 17:31:44 -- max datetime value expected, i.e. 2106-02-07 06:28:15
```

```sql
SELECT
    toDateTime64('2160-01-01 00:00:00', 0) AS a,
    toDateTime('2160-01-01 00:00:00') AS b,
    toDate(a),
    toDateTime(a)

Query id: a6085a36-4a41-467b-915d-b69638445754

Row 1:
──────
a:                                                  2160-01-01 00:00:00
b:                                                  2106-02-07 06:28:15
toDate(toDateTime64('2160-01-01 00:00:00', 0)):     1980-07-27 -- max malue, i.e. 2149-06-06  is expected
toDateTime(toDateTime64('2160-01-01 00:00:00', 0)): 2023-11-24 17:31:44 -- max value, i.e. 2106-02-07 06:28:15 is expected
```

P.S. It seems like fixing that may also be a fix for https://github.com/ClickHouse/ClickHouse/issues/39248 (which also can be fixed independently).

**Comment by filimonov** [comment]
More tests with negative timestamps:

```
SELECT
    toDate32('1930-01-01') AS a,
    toDate('1930-01-01') AS b,
    toDate(a),
    toDateTime(a),
    toDateTime(b)

Query id: 1b99647e-fa0b-45c2-bfa3-d1bfab4e7ab2

Row 1:
──────
a:                                  1930-01-01
b:                                  1970-01-01
toDate(toDate32('1930-01-01')):     2109-06-07 -- 1970-01-01!
toDateTime(toDate32('1930-01-01')): 2066-02-06 06:28:16  -- 1970-01-01 00:00:00!
toDateTime(toDate('1930-01-01')):   1970-01-01 00:00:00
```

**Comment by filimonov** [comment]
The same behavior is expected from toStartOf.. and other functions when 'cropping' the datatype happens.
https://github.com/ClickHouse/ClickHouse/issues/25284

**Comment by rschu1ze** [comment]
The fix was reverted as it caused a regression.

There was a question if we need narrowing at all or if the current behavior (undefined behavior for arguments outside the expected range) should not be accepted and documented (--> https://github.com/ClickHouse/ClickHouse/pull/40217#issuecomment-1287413708). Opinions?

Allow me to re-open so we don't forget about this.

@zvonand

**Comment by zvonand** [comment]
WDYT of reintroducing #40217, but with an option to use different workflows?

E.g. there would be a server setting like `truncate_datetime_overflow` which is 0 by default.
I fully understand that some may want to get predictable result instead of a semi-random value.

Or maybe a different approach: setting `allow_datetime_overflow`, 1 by default. If set to 0, an exception is thrown on overflow.

CC @rschu1ze

**Comment by UnamedRus** [comment]
There is setting:

```
decimal_check_overflow=0;
```

IMO, we should name date/time setting in similar way.

```
date(time)_check_overflow=0;
```

**Comment by rschu1ze** [comment]
Fixed as per #55696
